File: frontend/src/data/themeTokens.js
// ============================================
// ServerKit theme token whitelist + alias table + validators
// ============================================
// A ServerKit theme is DATA, not code (see docs/THEMING.md and plan 60): a map
// of canonical CSS custom-property names to values. This module is the single
// client-side source of truth for
//   - which token names a theme may set (the whitelist),
//   - how canonical names expand to the legacy `--bg-*/--text-*/--border-*`
//     aliases the stylesheet still reads (the alias table), and
//   - what a token's value may look like (the validators).
//
// KEEP IN SYNC with backend/app/services/theme_tokens.py — the backend runs the
// same whitelist + validation server-side on import and registry install.
//
// Accent tokens (--accent*, --accent-primary, …) are intentionally NOT in the
// whitelist: they are derived at runtime from a theme's single `accent` field
// by ThemeContext.applyAccentToDOM, so a workspace/user accent keeps precedence.

// Canonical writable tokens, grouped for the Theme Studio color-picker UI.
// Authors write these ~37 names; the alias table below fans each out to the
// legacy names so all ~60 theme-sensitive stylesheet variables get painted.
export const TOKEN_GROUPS = {
    surfaces: ['--bg-body', '--bg-sidebar', '--surface', '--surface-2', '--surface-3', '--surface-hover'],
    borders: ['--border', '--border-soft', '--border-strong'],
    text: ['--text', '--text-dim', '--text-faint', '--text-ghost'],
    semantic: ['--green', '--green-bg', '--amber', '--amber-bg', '--red', '--red-bg', '--cyan', '--cyan-bg', '--violet', '--violet-bg'],
    radius: ['--radius', '--radius-sm', '--radius-lg'],
    fonts: ['--sans', '--mono'],
    shadows: ['--shadow-sm', '--shadow-md', '--shadow-lg'],
    scrollbar: ['--scrollbar-track', '--scrollbar-thumb', '--scrollbar-thumb-hover'],
    code: ['--bg-code', '--text-code'],
    chrome: ['--grid-color'],
};

// Human labels for each group (Theme Studio section headers).
export const GROUP_LABELS = {
    surfaces: 'Surfaces',
    borders: 'Borders',
    text: 'Text',
    semantic: 'Semantic colors',
    radius: 'Radius',
    fonts: 'Fonts',
    shadows: 'Shadows',
    scrollbar: 'Scrollbar',
    code: 'Code / terminal',
    chrome: 'Chrome',
};

// Which validator applies to each group.
const GROUP_TYPE = {
    surfaces: 'color',
    borders: 'color',
    text: 'color',
    semantic: 'color',
    radius: 'length',
    fonts: 'font',
    shadows: 'shadow',
    scrollbar: 'color',
    code: 'color',
    chrome: 'color',
};

// Flat whitelist + token -> value-type index.
export const CANONICAL_TOKENS = Object.values(TOKEN_GROUPS).flat();

export const TOKEN_TYPE = (() => {
    const out = {};
    for (const [group, tokens] of Object.entries(TOKEN_GROUPS)) {
        for (const t of tokens) out[t] = GROUP_TYPE[group];
    }
    return out;
})();

// Canonical name -> legacy names that must receive the same value. Applying a
// canonical token also sets its aliases so pages still reading `$bg-card` etc.
// (which resolve to the legacy custom properties) track the skin.
export const ALIAS_MAP = {
    '--surface': ['--bg-card'],
    '--surface-2': ['--bg-elevated', '--bg-secondary'],
    '--surface-3': ['--bg-tertiary'],
    '--surface-hover': ['--bg-hover'],
    '--border': ['--border-default'],
    '--border-soft': ['--border-subtle'],
    '--border-strong': ['--border-active', '--border-hover'],
    '--text': ['--text-primary'],
    '--text-dim': ['--text-secondary'],
    '--text-faint': ['--text-tertiary'],
};

// Every property name a skin may ever set (canonical + aliases). Used to clear
// a previous skin cleanly before applying the next one — never touches accent.
export const ALL_SKIN_PROPERTIES = [
    ...CANONICAL_TOKENS,
    ...Object.values(ALIAS_MAP).flat(),
];

// --------------------------------------------
// Value validation
// --------------------------------------------
// Reject anything that could break out of a single CSS declaration or smuggle a
// network fetch / script. Even though we apply per-token via setProperty (which
// the browser sanitises), this keeps the stored/exported theme.json clean and
// makes the same rules enforceable server-side and in the registry CI.
const FORBIDDEN = /url\(|expression\(|javascript:|@import|[@;{}<>]|\/\*/i;

const HEX = /^#([0-9a-fA-F]{3,4}|[0-9a-fA-F]{6}|[0-9a-fA-F]{8})$/;
const FUNC_COLOR = /^(rgb|rgba|hsl|hsla)\(\s*[0-9.,%\s/]+\)$/i;
const NAMED_COLOR = /^[a-zA-Z]{3,24}$/; // transparent, currentColor, white, …
const LENGTH = /^-?(?:0|\d*\.?\d+)(px|rem|em|%|vh|vw)?$/;
const FONT_STACK = /^[a-zA-Z0-9 ,"'-]{1,200}$/;
const SHADOW = /^[0-9a-zA-Z.,%()#\s/-]{1,200}$/; // numbers, rgba(), inset, none

function isColor(v) {
    return HEX.test(v) || FUNC_COLOR.test(v) || NAMED_COLOR.test(v);
}

const VALIDATORS = {
    color: isColor,
    length: (v) => LENGTH.test(v),
    font: (v) => FONT_STACK.test(v),
    shadow: (v) => SHADOW.test(v) || v === 'none',
};

/**
 * Validate a single token value. Returns the trimmed value if the token is
 * whitelisted and the value passes its type check (and contains no forbidden
 * substring), otherwise null.
 */
export function validateTokenValue(token, value) {
    const type = TOKEN_TYPE[token];
    if (!type) return null;                       // not whitelisted
    if (typeof value !== 'string') return null;
    const v = value.trim();
    if (!v || v.length > 200) return null;
    if (FORBIDDEN.test(v)) return null;
    return VALIDATORS[type](v) ? v : null;
}

/**
 * Drop unknown / invalid entries from a raw token map, returning only the
 * whitelisted, validated canonical tokens. Unknown keys degrade gracefully
 * (forward-compat: a newer theme on an older panel just loses the new tokens).
 */
export function sanitizeTokens(raw) {
    const out = {};
    if (!raw || typeof raw !== 'object') return out;
    for (const [token, value] of Object.entries(raw)) {
        const clean = validateTokenValue(token, value);
        if (clean !== null) out[token] = clean;
    }
    return out;
}

/**
 * Expand a validated canonical token map to include its legacy aliases, ready
 * to hand to setProperty. Canonical values win over any alias collision.
 */
export function expandAliases(cleanTokens) {
    const out = { ...cleanTokens };
    for (const [token, value] of Object.entries(cleanTokens)) {
        for (const alias of ALIAS_MAP[token] || []) {
            if (!(alias in cleanTokens)) out[alias] = value;
        }
    }
    return out;
}

File: backend/app/services/theme_tokens.py
"""Canonical theme token whitelist + alias table + validators (server side).

A ServerKit theme is DATA, not code (plan 60): a map of canonical CSS
custom-property names to values. This module is the backend source of truth for
the same three things the panel enforces client-side in
``frontend/src/data/themeTokens.js`` — KEEP THE TWO IN SYNC:

  - which token names a theme may set (the whitelist),
  - how canonical names expand to the legacy ``--bg-*/--text-*/--border-*``
    aliases the stylesheet still reads, and
  - what a token's value may look like (the validators).

The backend runs this on ``POST /import`` and on registry install so a malicious
or malformed theme never reaches the DB — the client validation is a
convenience, this is the gate.
"""
import re

# Canonical writable tokens, grouped. Authors write these ~37 names; the alias
# table fans each out to the legacy names. Accent tokens are intentionally
# excluded — they are derived at runtime from a theme's single ``accent`` field.
TOKEN_GROUPS = {
    'surfaces': ['--bg-body', '--bg-sidebar', '--surface', '--surface-2', '--surface-3', '--surface-hover'],
    'borders': ['--border', '--border-soft', '--border-strong'],
    'text': ['--text', '--text-dim', '--text-faint', '--text-ghost'],
    'semantic': ['--green', '--green-bg', '--amber', '--amber-bg', '--red', '--red-bg', '--cyan', '--cyan-bg', '--violet', '--violet-bg'],
    'radius': ['--radius', '--radius-sm', '--radius-lg'],
    'fonts': ['--sans', '--mono'],
    'shadows': ['--shadow-sm', '--shadow-md', '--shadow-lg'],
    'scrollbar': ['--scrollbar-track', '--scrollbar-thumb', '--scrollbar-thumb-hover'],
    'code': ['--bg-code', '--text-code'],
    'chrome': ['--grid-color'],
}

_GROUP_TYPE = {
    'surfaces': 'color',
    'borders': 'color',
    'text': 'color',
    'semantic': 'color',
    'radius': 'length',
    'fonts': 'font',
    'shadows': 'shadow',
    'scrollbar': 'color',
    'code': 'color',
    'chrome': 'color',
}

CANONICAL_TOKENS = [t for group in TOKEN_GROUPS.values() for t in group]

TOKEN_TYPE = {t: _GROUP_TYPE[g] for g, tokens in TOKEN_GROUPS.items() for t in tokens}

# Canonical name -> legacy names that must receive the same value.
ALIAS_MAP = {
    '--surface': ['--bg-card'],
    '--surface-2': ['--bg-elevated', '--bg-secondary'],
    '--surface-3': ['--bg-tertiary'],
    '--surface-hover': ['--bg-hover'],
    '--border': ['--border-default'],
    '--border-soft': ['--border-subtle'],
    '--border-strong': ['--border-active', '--border-hover'],
    '--text': ['--text-primary'],
    '--text-dim': ['--text-secondary'],
    '--text-faint': ['--text-tertiary'],
}

VALID_BASES = ('dark', 'light')
VALID_SOURCES = ('registry', 'import', 'studio', 'bundled')

# --------------------------------------------
# Value validation
# --------------------------------------------
_FORBIDDEN = re.compile(r'url\(|expression\(|javascript:|@import|[@;{}<>]|/\*', re.IGNORECASE)
_HEX = re.compile(r'^#([0-9a-fA-F]{3,4}|[0-9a-fA-F]{6}|[0-9a-fA-F]{8})$')
_FUNC_COLOR = re.compile(r'^(rgb|rgba|hsl|hsla)\(\s*[0-9.,%\s/]+\)$', re.IGNORECASE)
_NAMED_COLOR = re.compile(r'^[a-zA-Z]{3,24}$')
_LENGTH = re.compile(r'^-?(?:0|\d*\.?\d+)(px|rem|em|%|vh|vw)?$')
_FONT_STACK = re.compile(r'^[a-zA-Z0-9 ,"\'-]{1,200}$')
_SHADOW = re.compile(r'^[0-9a-zA-Z.,%()#\s/-]{1,200}$')


def _is_color(v):
    return bool(_HEX.match(v) or _FUNC_COLOR.match(v) or _NAMED_COLOR.match(v))


_VALIDATORS = {
    'color': _is_color,
    'length': lambda v: bool(_LENGTH.match(v)),
    'font': lambda v: bool(_FONT_STACK.match(v)),
    'shadow': lambda v: v == 'none' or bool(_SHADOW.match(v)),
}


def validate_token_value(token, value):
    """Return the trimmed value if ``token`` is whitelisted and ``value`` passes
    its type check (and has no forbidden substring), else ``None``."""
    ttype = TOKEN_TYPE.get(token)
    if not ttype:
        return None
    if not isinstance(value, str):
        return None
    v = value.strip()
    if not v or len(v) > 200:
        return None
    if _FORBIDDEN.search(v):
        return None
    return v if _VALIDATORS[ttype](v) else None


def sanitize_tokens(raw):
    """Drop unknown / invalid entries, returning only whitelisted, validated
    canonical tokens. Unknown keys degrade gracefully (forward-compat)."""
    out = {}
    if not isinstance(raw, dict):
        return out
    for token, value in raw.items():
        clean = validate_token_value(token, value)
        if clean is not None:
            out[token] = clean
    return out


def expand_aliases(clean_tokens):
    """Expand a validated canonical map to include legacy aliases."""
    out = dict(clean_tokens)
    for token, value in clean_tokens.items():
        for alias in ALIAS_MAP.get(token, []):
            if alias not in clean_tokens:
                out[alias] = value
    return out


_SLUG_RE = re.compile(r'^[a-z0-9]+(?:-[a-z0-9]+)*$')


def validate_theme(raw):
    """Validate and normalise a raw theme dict (from import / registry / studio).

    Returns ``(theme, error)``: on success ``theme`` is a cleaned dict with a
    sanitised ``tokens`` map (invalid/unknown tokens dropped) and ``error`` is
    ``None``; on failure ``theme`` is ``None`` and ``error`` is a message.
    """
    if not isinstance(raw, dict):
        return None, 'Theme must be a JSON object'
    if raw.get('schema_version') not in (1, '1', None):
        return None, 'Unsupported schema_version (expected 1)'

    slug = raw.get('slug')
    if not isinstance(slug, str) or not _SLUG_RE.match(slug):
        return None, 'Invalid or missing slug (kebab-case, lowercase alphanumerics)'

    name = raw.get('name')
    if not isinstance(name, str) or not name.strip():
        return None, 'Missing theme name'

    base = raw.get('base', 'dark')
    if base not in VALID_BASES:
        return None, "base must be 'dark' or 'light'"

    raw_tokens = raw.get('tokens')
    if not isinstance(raw_tokens, dict):
        return None, 'Missing tokens object'

    tokens = {}
    for mode in ('dark', 'light'):
        if isinstance(raw_tokens.get(mode), dict):
            cleaned = sanitize_tokens(raw_tokens[mode])
            if cleaned:
                tokens[mode] = cleaned
    if not tokens:
        return None, 'Theme defines no valid tokens'

    accent = raw.get('accent')
    if accent is not None and validate_token_value('--surface', accent) is None:
        # accent is a plain color; reuse the color validator via a color token.
        accent = None

    preview = raw.get('preview')
    if (isinstance(preview, list) and len(preview) == 4
            and all(isinstance(c, str) for c in preview)):
        # Swatches are painted as inline CSS — every one must be a real color
        # (same gate as accent), never an url()/expression() payload.
        if any(validate_token_value('--surface', c) is None for c in preview):
            return None, 'preview swatches must be valid colors'
    else:
        # Derive a reasonable preview from the tokens if absent/malformed.
        primary = tokens.get('dark') or tokens.get('light') or {}
        preview = [
            primary.get('--bg-body', '#101218'),
            primary.get('--surface', '#161922'),
            accent or '#6d7cff',
            primary.get('--text', '#e9ebf0'),
        ]

    theme = {
        'schema_version': 1,
        'slug': slug,
        'name': name.strip(),
        'author': raw.get('author') if isinstance(raw.get('author'), str) else '',
        'version': raw.get('version') if isinstance(raw.get('version'), str) else '1.0.0',
        'description': raw.get('description') if isinstance(raw.get('description'), str) else '',
        'base': base,
        'tokens': tokens,
        'accent': accent,
        'preview': preview[:4],
    }
    return theme, None
